File: pipeline/reimpl/cobil00c.py
"""
Reimplementation of COBIL00C — CardDemo Bill Payment Screen.

CICS online program. Allows a user to pay their full account balance.
Steps:
  1. Enter account ID → displays current balance
  2. Confirm with 'Y' → creates payment transaction, zeroes balance

Navigation:
  ENTER → process (lookup then confirm)
  PF3   → return to calling program
  PF4   → clear screen
"""

from __future__ import annotations
from dataclasses import dataclass
from decimal import Decimal
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader, AccountRecord, TranRecord,
    CardXrefRecord, CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3, DFHPF4,
    get_db2_timestamp,
)


WS_PGMNAME = "COBIL00C"
WS_TRANID  = "CB00"
BILL_PAY_TYPE_CD   = "02"
BILL_PAY_CAT_CD    = 2
BILL_PAY_SOURCE    = "POS TERM"
BILL_PAY_DESC      = "BILL PAYMENT - ONLINE"
BILL_PAY_MERCHANT_ID   = 999999999
BILL_PAY_MERCHANT_NAME = "BILL PAYMENT"
BILL_PAY_MERCHANT_CITY = "N/A"
BILL_PAY_MERCHANT_ZIP  = "N/A"


@dataclass
class BillPayResult:
    acct_record: Optional[AccountRecord] = None
    tran_record: Optional[TranRecord] = None
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    success: bool = False
    xctl_program: str = ""
    return_to_prev: bool = False
    cleared: bool = False


class AccountRepository:
    def __init__(self, accounts: dict[int, AccountRecord]):
        self._accounts = accounts

    def find(self, acct_id: int) -> Optional[AccountRecord]:
        return self._accounts.get(acct_id)

    def rewrite(self, acct: AccountRecord) -> bool:
        if acct.acct_id not in self._accounts:
            return False
        self._accounts[acct.acct_id] = acct
        return True


class XrefRepository:
    def __init__(self, xrefs: dict[int, CardXrefRecord]):
        self._by_acct = xrefs

    def find_by_acct(self, acct_id: int) -> Optional[CardXrefRecord]:
        return self._by_acct.get(acct_id)


class TranRepository:
    def __init__(self, transactions: list[TranRecord]):
        self._trans = sorted(transactions, key=lambda t: t.tran_id)

    def next_id(self) -> str:
        if not self._trans:
            return "0000000000000001"
        last_id = self._trans[-1].tran_id.strip()
        try:
            return str(int(last_id) + 1).zfill(16)
        except ValueError:
            return "0000000000000001"

    def write(self, tran: TranRecord) -> bool:
        self._trans.append(tran)
        self._trans.sort(key=lambda t: t.tran_id)
        return True


def process_bill_pay(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    acct_id_input: str,
    confirm: str,
    acct_repo: AccountRepository,
    xref_repo: XrefRepository,
    tran_repo: TranRepository,
) -> BillPayResult:
    """Process bill payment screen — mirrors COBIL00C PROCEDURE DIVISION."""
    result = BillPayResult()

    if eibcalen == 0:
        result.return_to_prev = True
        result.xctl_program = "COSGN00C"
        return result

    result.commarea = commarea

    if commarea.cdemo_pgm_context == 0:
        commarea.cdemo_pgm_context = 1
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.commarea = commarea
        return result

    if eibaid in (DFHENTER, "ENTER"):
        return _process_enter(acct_id_input, confirm, acct_repo, xref_repo, tran_repo, commarea, result)
    elif eibaid in (DFHPF3, "PF3"):
        back = commarea.cdemo_from_program or "COMEN01C"
        result.xctl_program = back
        result.return_to_prev = True
        result.commarea = commarea
        return result
    elif eibaid in (DFHPF4, "PF4"):
        result.cleared = True
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.commarea = commarea
        return result
    else:
        result.error = True
        result.message = CCDA_MSG_INVALID_KEY
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result


def _process_enter(acct_id_input, confirm, acct_repo, xref_repo, tran_repo, commarea, result):
    if not acct_id_input or acct_id_input.strip() == "":
        result.error = True
        result.message = "Acct ID can NOT be empty..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    try:
        acct_id = int(acct_id_input.strip())
    except ValueError:
        result.error = True
        result.message = "Acct ID must be numeric..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    conf = confirm.strip().upper() if confirm else ""

    if conf in ("N",):
        result.cleared = True
        result.error = True
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.commarea = commarea
        return result
    elif conf not in ("Y", "", "\x00"):
        result.error = True
        result.message = "Invalid value. Valid values are (Y/N)..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    acct = acct_repo.find(acct_id)
    if acct is None:
        result.error = True
        result.message = f"Account {acct_id} not found..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    result.acct_record = acct

    if acct.acct_curr_bal <= Decimal("0.00"):
        result.error = True
        result.message = "You have nothing to pay..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    if conf == "Y":
        xref = xref_repo.find_by_acct(acct_id)
        card_num = xref.xref_card_num.strip() if xref else ""
        ts = get_db2_timestamp()

        tran = TranRecord(
            tran_id=tran_repo.next_id(),
            tran_type_cd=BILL_PAY_TYPE_CD,
            tran_cat_cd=BILL_PAY_CAT_CD,
            tran_source=BILL_PAY_SOURCE,
            tran_desc=BILL_PAY_DESC,
            tran_amt=acct.acct_curr_bal,
            tran_card_num=card_num,
            tran_merchant_id=BILL_PAY_MERCHANT_ID,
            tran_merchant_name=BILL_PAY_MERCHANT_NAME,
            tran_merchant_city=BILL_PAY_MERCHANT_CITY,
            tran_merchant_zip=BILL_PAY_MERCHANT_ZIP,
            tran_orig_ts=ts,
            tran_proc_ts=ts,
        )
        tran_repo.write(tran)
        acct.acct_curr_bal = acct.acct_curr_bal - tran.tran_amt
        acct_repo.rewrite(acct)

        result.success = True
        result.tran_record = tran
        result.message = "Bill payment successful."
    else:
        result.message = "Confirm to make a bill payment..."

    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message
    result.commarea = commarea
    return result

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cobil00c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.util.*;

/**
 * Java reimplementation of COBIL00C — CardDemo Bill Payment Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/cobil00c.py}. Python is the source of truth.
 *
 * <p>Structure: standalone {@link #processBillPay} method + thin
 * {@link #runVector} adapter, matching the Python's
 * {@code process_bill_pay} / {@code run_vector} split.
 */
public class Cobil00c implements ProgramRunner {

    private static final String CCDA_MSG_INVALID_KEY = "Invalid key pressed.                    ";

    // ── Inner data types ────────────────────────────────────────────────────

    static class AccountRecord {
        int acctId;
        String acctActiveStatus;
        BigDecimal acctCurrBal;
        BigDecimal acctCreditLimit;

        AccountRecord(int acctId, String acctActiveStatus, BigDecimal acctCurrBal, BigDecimal acctCreditLimit) {
            this.acctId = acctId;
            this.acctActiveStatus = acctActiveStatus;
            this.acctCurrBal = acctCurrBal;
            this.acctCreditLimit = acctCreditLimit;
        }

        AccountRecord copy() {
            return new AccountRecord(acctId, acctActiveStatus,
                    new BigDecimal(acctCurrBal.toPlainString()),
                    new BigDecimal(acctCreditLimit.toPlainString()));
        }
    }

    static class CardXrefRecord {
        String xrefCardNum;
        int xrefCustId;
        int xrefAcctId;

        CardXrefRecord(String xrefCardNum, int xrefCustId, int xrefAcctId) {
            this.xrefCardNum = xrefCardNum;
            this.xrefCustId = xrefCustId;
            this.xrefAcctId = xrefAcctId;
        }
    }

    static class TranRecord {
        String tranId;
        String tranTypeCd;
        int tranCatCd;
        String tranSource;
        String tranDesc;
        BigDecimal tranAmt;
        String tranCardNum;

        TranRecord(String tranId, String tranTypeCd, int tranCatCd, String tranSource,
                   String tranDesc, BigDecimal tranAmt, String tranCardNum) {
            this.tranId = tranId;
            this.tranTypeCd = tranTypeCd;
            this.tranCatCd = tranCatCd;
            this.tranSource = tranSource;
            this.tranDesc = tranDesc;
            this.tranAmt = tranAmt;
            this.tranCardNum = tranCardNum;
        }
    }

    static class BillPayResult {
        AccountRecord acctRecord = null;
        TranRecord tranRecord = null;
        String message = "";
        boolean error = false;
        boolean success = false;
        String xctlProgram = "";
        boolean returnToPrev = false;
        boolean cleared = false;
    }

    // ── Repositories ────────────────────────────────────────────────────────

    static class AccountRepository {
        private final Map<Integer, AccountRecord> accounts;

        AccountRepository(Map<Integer, AccountRecord> accounts) {
            this.accounts = new HashMap<>(accounts);
        }

        AccountRecord find(int acctId) {
            return accounts.get(acctId);
        }

        boolean rewrite(AccountRecord acct) {
            if (!accounts.containsKey(acct.acctId)) return false;
            accounts.put(acct.acctId, acct);
            return true;
        }
    }

    static class XrefRepository {
        private final Map<Integer, CardXrefRecord> byAcct;

        XrefRepository(Map<Integer, CardXrefRecord> byAcct) {
            this.byAcct = new HashMap<>(byAcct);
        }

        CardXrefRecord findByAcct(int acctId) {
            return byAcct.get(acctId);
        }
    }

    static class TranRepository {
        private final List<TranRecord> trans;

        TranRepository(List<TranRecord> trans) {
            this.trans = new ArrayList<>(trans);
            this.trans.sort(Comparator.comparing(t -> t.tranId));
        }

        String nextId() {
            if (trans.isEmpty()) return "0000000000000001";
            String lastId = trans.get(trans.size() - 1).tranId.trim();
            try {
                long next = Long.parseLong(lastId) + 1;
                return String.format("%016d", next);
            } catch (NumberFormatException e) {
                return "0000000000000001";
            }
        }

        boolean write(TranRecord tran) {
            trans.add(tran);
            trans.sort(Comparator.comparing(t -> t.tranId));
            return true;
        }
    }

    // ── Standalone business method ──────────────────────────────────────────

    /**
     * Process bill payment screen — mirrors {@code process_bill_pay} in
     * {@code pipeline/reimpl/cobil00c.py}.
     */
    BillPayResult processBillPay(
            int eibcalen, String eibaid, int pgmContext,
            String acctIdInput, String confirm,
            AccountRepository acctRepo, XrefRepository xrefRepo, TranRepository tranRepo) {

        BillPayResult result = new BillPayResult();

        if (eibcalen == 0) {
            result.returnToPrev = true;
            result.xctlProgram = "COSGN00C";
            return result;
        }

        if (pgmContext == 0) {
            // First re-entry — show blank bill pay screen
            return result;
        }

        if ("ENTER".equals(eibaid)) {
            return processEnter(acctIdInput, confirm, acctRepo, xrefRepo, tranRepo, result);
        } else if ("PF3".equals(eibaid)) {
            result.xctlProgram = "COMEN01C";
            result.returnToPrev = true;
            return result;
        } else if ("PF4".equals(eibaid)) {
            result.cleared = true;
            return result;
        } else {
            result.error = true;
            result.message = CCDA_MSG_INVALID_KEY;
            return result;
        }
    }

    private BillPayResult processEnter(
            String acctIdInput, String confirm,
            AccountRepository acctRepo, XrefRepository xrefRepo, TranRepository tranRepo,
            BillPayResult result) {

        if (acctIdInput == null || acctIdInput.trim().isEmpty()) {
            result.error = true;
            result.message = "Acct ID can NOT be empty...";
            return result;
        }

        int acctId;
        try {
            acctId = Integer.parseInt(acctIdInput.trim());
        } catch (NumberFormatException e) {
            result.error = true;
            result.message = "Acct ID must be numeric...";
            return result;
        }

        String conf = confirm != null ? confirm.trim().toUpperCase() : "";

        if ("N".equals(conf)) {
            result.cleared = true;
            result.error = true;
            return result;
        } else if (!conf.isEmpty() && !"\u0000".equals(conf) && !"Y".equals(conf)) {
            result.error = true;
            result.message = "Invalid value. Valid values are (Y/N)...";
            return result;
        }

        AccountRecord acct = acctRepo.find(acctId);
        if (acct == null) {
            result.error = true;
            result.message = "Account " + acctId + " not found...";
            return result;
        }

        result.acctRecord = acct;

        if (acct.acctCurrBal.compareTo(BigDecimal.ZERO) <= 0) {
            result.error = true;
            result.message = "You have nothing to pay...";
            return result;
        }

        if ("Y".equals(conf)) {
            CardXrefRecord xref = xrefRepo.findByAcct(acctId);
            String cardNum = xref != null ? xref.xrefCardNum.trim() : "";

            TranRecord tran = new TranRecord(
                    tranRepo.nextId(), "02", 2, "POS TERM",
                    "BILL PAYMENT - ONLINE", acct.acctCurrBal, cardNum);
            tranRepo.write(tran);
            acct.acctCurrBal = acct.acctCurrBal.subtract(tran.tranAmt);
            acctRepo.rewrite(acct);

            result.success = true;
            result.tranRecord = tran;
            result.message = "Bill payment successful.";
        } else {
            result.message = "Confirm to make a bill payment...";
        }

        return result;
    }

    // ── runVector adapter ───────────────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "PAY_SUCCESS");

        // Seed data
        Map<Integer, AccountRecord> seedAccounts = new HashMap<>();
        seedAccounts.put(10000001, new AccountRecord(10000001, "Y",
                new BigDecimal("1500.00"), new BigDecimal("5000.00")));
        seedAccounts.put(10000002, new AccountRecord(10000002, "Y",
                BigDecimal.ZERO, new BigDecimal("3000.00")));

        AccountRepository acctRepo = new AccountRepository(seedAccounts);

        Map<Integer, CardXrefRecord> seedXrefs = new HashMap<>();
        seedXrefs.put(10000001, new CardXrefRecord("4111111111111111", 1, 10000001));
        XrefRepository xrefRepo = new XrefRepository(seedXrefs);

        List<TranRecord> seedTrans = new ArrayList<>();
        seedTrans.add(new TranRecord("0000000000000001", "01", 1, "ONLINE",
                "PURCHASE", new BigDecimal("50.00"), "4111111111111111"));
        TranRepository tranRepo = new TranRepository(seedTrans);

        BillPayResult result;

        switch (scenario) {
            case "PAY_SUCCESS":
                result = processBillPay(100, "ENTER", 1,
                        "10000001", "Y", acctRepo, xrefRepo, tranRepo);
                break;
            case "ACCT_NOT_FOUND":
                result = processBillPay(100, "ENTER", 1,
                        "99999999", "", acctRepo, xrefRepo, tranRepo);
                break;
            case "ZERO_BALANCE":
                result = processBillPay(100, "ENTER", 1,
                        "10000002", "", acctRepo, xrefRepo, tranRepo);
                break;
            case "EMPTY_ACCT_ID":
                result = processBillPay(100, "ENTER", 1,
                        "", "", acctRepo, xrefRepo, tranRepo);
                break;
            case "CONFIRM_NO":
                result = processBillPay(100, "ENTER", 1,
                        "10000001", "N", acctRepo, xrefRepo, tranRepo);
                break;
            case "INVALID_KEY":
                result = processBillPay(100, "PF9", 1,
                        "", "", acctRepo, xrefRepo, tranRepo);
                break;
            default:
                result = processBillPay(100, "ENTER", 1,
                        "10000001", "Y", acctRepo, xrefRepo, tranRepo);
                break;
        }

        String tranId = "";
        String tranAmt = "";
        if (result.tranRecord != null) {
            tranId = result.tranRecord.tranId;
            tranAmt = result.tranRecord.tranAmt.toPlainString();
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("SUCCESS", result.success ? "Y" : "N");
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("TRAN_ID", tranId);
        out.put("TRAN_AMT", tranAmt);
        out.put("XCTL_PROGRAM", result.xctlProgram != null ? result.xctlProgram : "");
        return out;
    }
}
